### PR TITLE
More explicit error when sudo requires password for actions

### PR DESCRIPTION
**Pull request by lakshmi-kannan**
```
(virtualenv)~/stanley git:linux_pack_improvements ❯❯❯ st2 run linux.lsof_pids hosts=localhost pids=20430 username=stanley                                                                                                       ✭ ✱ ◼
.
+-----------------+--------------------------------------------------------------+
| Property        | Value                                                        |
+-----------------+--------------------------------------------------------------+
| id              | 54f0fcd10640fd02810515cc                                     |
| action.ref      | linux.lsof_pids                                              |
| context.user    | lakshmi                                                      |
| parameters      | {                                                            |
|                 |     "username": "stanley",                                   |
|                 |     "pids": "20430",                                         |
|                 |     "hosts": "localhost"                                     |
|                 | }                                                            |
| status          | failed                                                       |
| start_timestamp | Fri, 27 Feb 2015 23:25:05 UTC                                |
| end_timestamp   | Fri, 27 Feb 2015 23:25:06 UTC                                |
| result          | {                                                            |
|                 |     "localhost": {                                           |
|                 |         "failed": true,                                      |
|                 |         "traceback": "  File "/home/lakshmi/stanley/st2commo |
|                 | n/st2common/models/system/action.py", line 376, in _sudo     |
|                 |     output = sudo(self.command, combine_stderr=False,        |
|                 | pty=True, quiet=True)                                        |
|                 |   File "/home/lakshmi/stanley/virtualenv/local/lib/python2.7 |
|                 | /site-packages/fabric/network.py", line 639, in              |
|                 | host_prompting_wrapper                                       |
|                 |     return func(*args, **kwargs)                             |
|                 |   File "/home/lakshmi/stanley/virtualenv/local/lib/python2.7 |
|                 | /site-packages/fabric/operations.py", line 1095, in sudo     |
|                 |     stderr=stderr, timeout=timeout,                          |
|                 | shell_escape=shell_escape,                                   |
|                 |   File "/home/lakshmi/stanley/virtualenv/local/lib/python2.7 |
|                 | /site-packages/fabric/operations.py", line 909, in           |
|                 | _run_command                                                 |
|                 |     channel=default_channel(), command=wrapped_command,      |
|                 | pty=pty,                                                     |
|                 |   File "/home/lakshmi/stanley/virtualenv/local/lib/python2.7 |
|                 | /site-packages/fabric/state.py", line 390, in                |
|                 | default_channel                                              |
|                 |     chan = _open_session()                                   |
|                 |   File "/home/lakshmi/stanley/virtualenv/local/lib/python2.7 |
|                 | /site-packages/fabric/state.py", line 382, in _open_session  |
|                 |     return                                                   |
|                 | connections[env.host_string].get_transport().open_session()  |
|                 |   File "/home/lakshmi/stanley/virtualenv/local/lib/python2.7 |
|                 | /site-packages/fabric/network.py", line 151, in __getitem__  |
|                 |     self.connect(key)                                        |
|                 |   File "/home/lakshmi/stanley/virtualenv/local/lib/python2.7 |
|                 | /site-packages/fabric/network.py", line 143, in connect      |
|                 |     self[key] = connect(user, host, port, cache=self)        |
|                 |   File "/home/lakshmi/stanley/virtualenv/local/lib/python2.7 |
|                 | /site-packages/fabric/network.py", line 523, in connect      |
|                 |     password = prompt_for_password(text)                     |
|                 |   File "/home/lakshmi/stanley/virtualenv/local/lib/python2.7 |
|                 | /site-packages/fabric/network.py", line 596, in              |
|                 | prompt_for_password                                          |
|                 |     handle_prompt_abort("a connection or sudo password")     |
|                 |   File "/home/lakshmi/stanley/virtualenv/local/lib/python2.7 |
|                 | /site-packages/fabric/utils.py", line 154, in                |
|                 | handle_prompt_abort                                          |
|                 |     abort(reason % "input would be ambiguous in parallel     |
|                 | mode")                                                       |
|                 |   File "/home/lakshmi/stanley/virtualenv/local/lib/python2.7 |
|                 | /site-packages/fabric/utils.py", line 32, in abort           |
|                 |     raise env.abort_exception(msg)                           |
|                 | ",                                                           |
|                 |         "succeeded": false,                                  |
|                 |         "error": "Passwordless sudo needs to be setup for    |
|                 | user: stanley"                                               |
|                 |     }                                                        |
|                 | }                                                            |
+-----------------+--------------------------------------------------------------+
(virtualenv)~/stanley git:linux_pack_improvements ❯❯❯
```

### CODE REVIEW

**Comment by manasdk** [comment]
@lakshmi-kannan Can you paste the result of 
st2 execution get 54f0fcd10640fd02810515cc

**Comment by lakshmi-kannan** [comment]
```
(virtualenv)~/stanley git:passwordless_sudo_warning ❯❯❯ st2 execution get 54f0fcd10640fd02810515cc                                                                                                                                ✭ ◼
status: failed
result:
{
    "localhost": {
        "failed": true,
        "traceback": "  File "/home/lakshmi/stanley/st2common/st2common/models/system/action.py", line 376, in _sudo
    output = sudo(self.command, combine_stderr=False, pty=True, quiet=True)
  File "/home/lakshmi/stanley/virtualenv/local/lib/python2.7/site-packages/fabric/network.py", line 639, in host_prompting_wrapper
    return func(*args, **kwargs)
  File "/home/lakshmi/stanley/virtualenv/local/lib/python2.7/site-packages/fabric/operations.py", line 1095, in sudo
    stderr=stderr, timeout=timeout, shell_escape=shell_escape,
  File "/home/lakshmi/stanley/virtualenv/local/lib/python2.7/site-packages/fabric/operations.py", line 909, in _run_command
    channel=default_channel(), command=wrapped_command, pty=pty,
  File "/home/lakshmi/stanley/virtualenv/local/lib/python2.7/site-packages/fabric/state.py", line 390, in default_channel
    chan = _open_session()
  File "/home/lakshmi/stanley/virtualenv/local/lib/python2.7/site-packages/fabric/state.py", line 382, in _open_session
    return connections[env.host_string].get_transport().open_session()
  File "/home/lakshmi/stanley/virtualenv/local/lib/python2.7/site-packages/fabric/network.py", line 151, in __getitem__
    self.connect(key)
  File "/home/lakshmi/stanley/virtualenv/local/lib/python2.7/site-packages/fabric/network.py", line 143, in connect
    self[key] = connect(user, host, port, cache=self)
  File "/home/lakshmi/stanley/virtualenv/local/lib/python2.7/site-packages/fabric/network.py", line 523, in connect
    password = prompt_for_password(text)
  File "/home/lakshmi/stanley/virtualenv/local/lib/python2.7/site-packages/fabric/network.py", line 596, in prompt_for_password
    handle_prompt_abort("a connection or sudo password")
  File "/home/lakshmi/stanley/virtualenv/local/lib/python2.7/site-packages/fabric/utils.py", line 154, in handle_prompt_abort
    abort(reason % "input would be ambiguous in parallel mode")
  File "/home/lakshmi/stanley/virtualenv/local/lib/python2.7/site-packages/fabric/utils.py", line 32, in abort
    raise env.abort_exception(msg)
",
        "succeeded": false,
        "error": "Passwordless sudo needs to be setup for user: stanley"
    }
}
(virtualenv)~/stanley git:passwordless_sudo_warning ❯❯❯                                                                                                                                                                           ✭ ◼
	st2client/st2client/commands/outpu
```

**Comment by manasdk** [comment]
stupid traceback takes away from the message. Its just noise in most cases especially if there is a useful message.

Any thats a separate problem.

File: st2common/st2common/models/system/action.py [hunk 1]
```diff
@@ -102,6 +102,9 @@ def _get_error_result(self):
         result = {}
         result['failed'] = True
         result['succeeded'] = False
+        exc_value = str(exc_value)
+        if 'sudo password' in exc_value:
```
Comment (manasdk) [inline st2common/st2common/models/system/action.py]: :(

**Comment by manasdk** [comment]
oh well ...if this is the only way.

+1

**Comment by lakshmi-kannan** [comment]
I couldn't think of a way that works across distros. I'll merge it for now. If we have a better solution, we can fix later.

**Comment by Kami** [comment]
Yeah, I was thinking if we can maybe catch a more specific fabric error message, but this should work as well.

**Comment by Kami** [comment]
Actually, to make false positives less likely, might be better to look for `a connection or sudo password`, right?

**Comment by lakshmi-kannan** [comment]
>Actually, to make false positives less likely, might be better to look for a connection or sudo password, right?

I think the current version is fine.

**Comment by Kami** [comment]
Btw, what's the `type(exc)` here? Just slightly worried about false positives.

**Comment by lakshmi-kannan** [comment]
It's a FabricExecutionException. I checked that.

**Comment by Kami** [comment]
Ah, ok, to be on the safe side we should check for that type and the text :)

**Comment by lakshmi-kannan** [comment]
>Ah, ok, to be on the safe side we should check for that type and the text :)

Fair enough. Let me do that.

**Comment by lakshmi-kannan** [comment]
@Kami Am I good to go?

**Comment by Kami** [comment]
+1

**Comment by Kami** [comment]
It turns out fabric returns the same error message (`Needed to prompt for a connection or sudo password (host: x4, but input would be ambiguous in parallel mode)` if it can't authenticate to the server (e.g. invalid username / password / key) which results in some confusing false positives.

I'm looking into it and trying to figure out if we can differentiate between those errors. If we can't we should probably revert this change since it makes it confusing - in case of an authentication failure, this has nothing to do with a passwordless sudo, it just makes it very confusing for the end user and harder to debug.

**Pull request merged**